keps/sig-node/3702-ima-namespace-support/README.md

```diff
@@ -0,0 +1,266 @@
+
+* Detect illicit activity
+* Detect unintended changes
+* Verify the status and health of the system
+* Comply with access rules
+
+This can be achieved using IMA (Integrity Measurement Architecture) and EVM (Extended Verification Module), which can use a TPM chip as Hardware Root of Trust for high security environments.
+
+### Goals
+
+* Check integrity of regular files inside containers, and hence check POD integrity
+* Using remote attestation mechanism, check the integrity of a given POD or all PODS in the cluster deployed with IMA namespace enabled periodically
+* Alert about corrupted or compromised pods
+
+### Non-Goals
+.
+  
+
+## Proposal
+<!-- We propose to enable IMA linux namespaces in pods.
+Since IMA namespaces can be created when a container is launched, we can provide transparent integrity verification on any linux container.
+IMA and EVM can use a TPM chip as a hardware root of trust. Hence we can verify images against a set of golden hash values, as well as avoiding any further changes to the overlayfs to intercept calls and check the integrity of files. -->
+
+### User Stories (Optional)
+
+#### Story 1
+As a cluster admin, I want to detect undesired file changes, so that I can take out pods that have been compromised.
+
+An intruder perform some malicious changes inside a certain pod's files. The system should be able to detect those changed and alert about the inconsistent pod. The remote attestation framework could keep this pod running or make a copy (for forensic analysis) and delete it.
+
+#### Story 2
+As a cluster admin, I want to deploy a only proven and certified pods, so that I can comply with internal policies as well as security regulations.
+
+Let's say that we are working in a high security environment where only approved images can be deployed. In this scenario we can make sure that the pod deploy used an imaged that hasn't been tampered.
+
+#### Story 3
+As a cluster admin, I want to deny access to certain files inside the pod, so that a potential intruder can't access sensitive information.
+
+In some cases, we should not even allow root to modify certain files inside the container.
+
+### Notes/Constraints/Caveats (Optional)
+We need to enable IMA in the kernel and container runtimes (runC, CRI-O, docker, containerd, etc.). 
+
+### Risks and Mitigations
+
+  
+
+## Design Details
+  
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+In order for the this feature to work, the nodes where the pods will be deploy should have IMA enabled and a recent kernel that supports IMA namespaces (WIP, it should be merged soon). The nodes should also have a TPM chip. We could use nodAffinity of labels and annotations in nodes in order to select where to deploy the pods.
```

Comment:
@haircommander I think this would be yet again a "feature" that should be advertised by the runtime to kubelet , à la https://github.com/kubernetes/kubernetes/issues/99808. From this perspective QoS resources would be lucrative as they would possibly give you all required functionality on K8s side out-of the box: automatic per-node discovery, dynamic updates (runtime re-configuration), visibility in node status, defaults, scheduler support etc...